packages/venia-concept/src/components/Navigation/categoryTree.js

```diff
@@ -21,45 +24,81 @@ class Tree extends Component {
         ),
         onNavigate: func,
         rootNodeId: number.isRequired,
-        updateRootNodeId: func.isRequired
+        updateRootNodeId: func.isRequired,
+        currentId: number.isRequired
     };
 
     get leaves() {
-        const { nodes, onNavigate, rootNodeId, updateRootNodeId } = this.props;
-        const { childrenData: childNodeIds } = nodes[rootNodeId];
+        const {
+            classes,
+            onNavigate,
+            rootNodeId,
+            updateRootNodeId,
+            currentId
+        } = this.props;
 
-        const leaves = childNodeIds.reduce((elements, id) => {
-            const childNode = nodes[id];
-            const { childrenData, position } = childNode;
-            const isLeaf = !childrenData.length;
-            const elementProps = { nodeId: id, nodes };
+        return rootNodeId ? (
+            <Query query={navigationMenu} variables={{ id: rootNodeId }}>
```

Comment:
I would consider moving this query up one level, into `Navigation`.

Reply:
Jimmy's right! I also notice that you called the class `Tree`, rather than `CategoryTree`. It does seem like this contains a lot of generic tree logic, which might be applicable to more than just categories. Should it perhaps become a generic navigation tree which takes a render prop? That would make it even easier to detach the query from it, since the propTypes would also be detached.